clang loop optimizer pragma warns on rasterio.cpp.

### What is the bug?

[rasterio.cpp](https://github.com/OSGeo/gdal/blob/493ed18ef8ad85813501573043bb9e1053d0f47f/gcore/rasterio.cpp#L5380C2-L5380C14) triggers a complaint with UBSAN mode with recent clang (like in the last couple weeks), but as this is a custom environment, I don't know if this happens with older clangs. I've seen `pass-failed` happen a few years ago with non-gdal code.

```
#if defined(__clang__)
#pragma clang loop vectorize(disable)
#endif
```

``` 
gcore/rasterio.cpp:5954:5: error: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Werror,-Wpass-failed=transform-warning]
 5954 |     for (size_t i = 0; i < nIters; ++i)
      |     ^
gcore/rasterio.cpp:6049:5: error: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Werror,-Wpass-failed=transform-warning]
 6049 |     for (size_t i = 0; i < nIters; ++i)
      |     ^
2 errors generated.
```

Not sure why the loop doesn't unroll.

### Steps to reproduce the issue

I don't have great details on how to reproduce. Sorry.

### Versions and provenance

Hit this GDAL from head on 2025-01-10.

### Additional context

Not sure the best way to disable this in the GDAL code base and is that reasonable to do? Locally, I set `-Wno-pass-failed` for now and then the UBSAN run goes just fine.

I looked at CPL_NOSANITIZE_UNSIGNED_INT_OVERFLOW and maybe this is correct `__attribute__((no_sanitize_undefined))`?

I will investigate more.

--- a/gcore/rasterio.cpp
+++ b/gcore/rasterio.cpp
@@ -5944,6 +5944,11 @@ __attribute__((optimize("tree-vectorize")))
 #if defined(__GNUC__)
 __attribute__((noinline))
 #endif
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+// clang++ -O2 -fsanitize=undefined fails to vectorize, ignore that warning
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wpass-failed"
+#endif
 static void
 GDALInterleave2Byte(const uint8_t *CPL_RESTRICT pSrc,
                     uint8_t *CPL_RESTRICT pDst, size_t nIters)
@@ -5957,6 +5962,9 @@ GDALInterleave2Byte(const uint8_t *CPL_RESTRICT pSrc,
         pDst[2 * i + 1] = pSrc[i + 1 * nIters];
     }
 }
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+#pragma clang diagnostic pop
+#endif
 
 #endif
 
@@ -6039,6 +6047,11 @@ __attribute__((optimize("tree-vectorize")))
 #if defined(__GNUC__)
 __attribute__((noinline))
 #endif
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+// clang++ -O2 -fsanitize=undefined fails to vectorize, ignore that warning
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wpass-failed"
+#endif
 static void
 GDALInterleave4Byte(const uint8_t *CPL_RESTRICT pSrc,
                     uint8_t *CPL_RESTRICT pDst, size_t nIters)
@@ -6054,6 +6067,9 @@ GDALInterleave4Byte(const uint8_t *CPL_RESTRICT pSrc,
         pDst[4 * i + 3] = pSrc[i + 3 * nIters];
     }
 }
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+#pragma clang diagnostic pop
+#endif
 
 #endif
 

--- a/gcore/rasterio_ssse3.cpp
+++ b/gcore/rasterio_ssse3.cpp
@@ -420,6 +420,11 @@ __attribute__((optimize("tree-vectorize")))
 #if defined(__GNUC__)
 __attribute__((noinline))
 #endif
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+// clang++ -O2 -fsanitize=undefined fails to vectorize, ignore that warning
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wpass-failed"
+#endif
 static void
 GDALInterleave3Byte_SSSE3(const uint8_t *CPL_RESTRICT pSrc,
                           uint8_t *CPL_RESTRICT pDst, size_t nIters)
@@ -434,6 +439,9 @@ GDALInterleave3Byte_SSSE3(const uint8_t *CPL_RESTRICT pSrc,
         pDst[3 * i + 2] = pSrc[i + 2 * nIters];
     }
 }
+#if defined(__clang__) && !defined(__INTEL_CLANG_COMPILER)
+#pragma clang diagnostic pop
+#endif
 
 #endif
 

